sdk/python/kfp/dsl/for_loop.py

```diff
@@ -172,26 +192,82 @@ def from_raw_items(
+        return LoopParameterArgument(
             items=raw_items,
             name_code=name_code,
             channel_type=type(raw_items[0]).__name__,
         )
 
-    @classmethod
-    def name_is_loop_argument(cls, name: str) -> bool:
-        """Returns True if the given channel name looks like a loop argument.
 
-        Either it came from a withItems loop item or withParams loop
-        item.
+class LoopArtifactArgument(pipeline_channel.PipelineArtifactChannel):
+    """Represents the artifact arguments that are looped over in a ParallelFor
+    loop.
+
+    The class shouldn't be instantiated by the end user, rather it is
+    created automatically by a ParallelFor ops group.
+
+    To create a LoopArtifactArgument instance, use the factory method::
+
+        LoopArtifactArgument.from_pipeline_channel(...)
+
+
+    Attributes:
+        pipeline_channel: The PipelineChannel object this
+            LoopArtifactArgument is associated to.
+    """
+
+    def __init__(
+        self,
+        items: pipeline_channel.PipelineChannel,
+        name_code: Optional[str] = None,
+        name_override: Optional[str] = None,
+        **kwargs,
+    ):
+        """Initializes a LoopArtifactArgument object.
+
+        Args:
+            items: The PipelineChannel object this LoopArtifactArgument is
+                associated to.
+            name_code: A unique code used to identify these loop arguments.
+                Should match the code for the ParallelFor ops_group which created
+                these LoopArtifactArguments. This prevents parameter name collisions.
+            name_override: The override name for PipelineChannel.
+            **kwargs: Any other keyword arguments passed down to PipelineChannel.
         """
-        return  ('-' + cls.LOOP_ITEM_NAME_BASE) in name \
-          or (cls.LOOP_ITEM_PARAM_NAME_BASE + '-') in name
+        if (name_code is None) == (name_override is None):
+            raise ValueError(
+                'Expect one and only one of `name_code` and `name_override` to '
+                'be specified.')
+
+        if name_override is None:
+            super().__init__(name=_make_name(name_code), **kwargs)
+        else:
+            super().__init__(name=name_override, **kwargs)
+
+        self.items_or_pipeline_channel = items
+        self.is_with_items_loop_argument = not isinstance(
+            items, pipeline_channel.PipelineArtifactChannel)
+
+    @classmethod
+    def from_pipeline_channel(
+        cls,
+        channel: pipeline_channel.PipelineChannel,
+        is_artifact_list: bool = False,
+    ) -> 'LoopArtifactArgument':
+        """Creates a LooArtifactArgument object from a PipelineChannel
+        object."""
+        return LoopArtifactArgument(
+            items=channel,
+            name_override=channel.name + '-' + LOOP_ITEM_NAME_BASE,
+            task_name=channel.task_name,
+            channel_type=_get_loop_item_type(channel.channel_type) or 'String',
```

Comment:
See other comments about getting artifact types from the channel. I think there may be an easier way to do this without `_get_loop_item_type` needing to know how to handle both parameters and artifacts and perform additional string parsing.

Reply:
Simplified this, now `_get_loop_item_type` only handles parameter type checking.